src/test/java/org/cloudfoundry/promregator/auth/OAuth2XSUAAEnricherTest.java

```diff
@@ -47,51 +55,121 @@ public static void cleanupEnvironment() {
 
 	@Test
 	void testAppropriateJWTCall() {
-		this.ams.getOauthTokenHandler().setResponse(this.oAuthServerResponse);
-		
-		OAuth2XSUAAAuthenticationConfiguration authenticatorConfig = new OAuth2XSUAAAuthenticationConfiguration();
-		authenticatorConfig.setClient_id("client_id");
-		authenticatorConfig.setClient_secret("client_secret");
-		authenticatorConfig.setTokenServiceURL("http://localhost:9001/oauth/token");
-		
-		OAuth2XSUAAEnricher subject = new OAuth2XSUAAEnricher(authenticatorConfig);
-		
+		this.ams.getOauthTokenHandler().setStatus(200);
+		this.ams.getOauthTokenHandler().setResponse(this.oAuthServer200Response);
+
+		OAuth2XSUAAEnricher subject = new OAuth2XSUAAEnricher(getConfig());
+
 		HttpGet mockGet = Mockito.mock(HttpGet.class);
 		Mockito.when(mockGet.getURI()).thenAnswer((Answer<URI>) invocation -> new URI("http://localhost/target"));
-		
+
 		subject.enrichWithAuthentication(mockGet);
 
 		verify(mockGet).setHeader("Authorization", "Bearer someAccessToken");
 	}
 
+	@Test
+	void testBadJWTCall() {
+		this.ams.getOauthTokenHandler().setStatus(401);
+		this.ams.getOauthTokenHandler().setResponse(this.oAuthServer401Response);
+
+		/* OAuth2XSUAAEnricher subject =  */ new OAuth2XSUAAEnricher(getConfig());
+
+		// Here I would like to continue with enrichWithAuthentication but the second call does not
+		//hit the mock server apparently. Doen't know why ...
```

Comment:
We should check this once again...

Reply:
I sent some time on this issue today without being able to improve this.

When we continue with the test like:

```
		this.ams.getOauthTokenHandler().setStatus(401);
		this.ams.getOauthTokenHandler().setResponse(this.oAuthServer401Response);

		OAuth2XSUAAEnricher subject = new OAuth2XSUAAEnricher(getConfig());

		HttpGet mockGet = Mockito.mock(HttpGet.class);
		Mockito.when(mockGet.getURI()).thenAnswer((Answer<URI>) invocation -> new URI("http://127.0.0.1/target"));
		
		subject.enrichWithAuthentication(mockGet);

```

Does also not work when using a real `httpGet`. This does not make a difference anyway since that get requests gets only prepared, but never fires (... we are adding only the auth header).

There is a second request triggered, but that never hits the `DefaultOAuthHttpHandler` as we can see since the log line `log.debug("Request was received at handler");` is not issued to the log.

When we trigger a thread dump we get

```"main" #1 prio=5 os_prio=31 tid=0x00007ff35f80d800 nid=0xe03 runnable [0x0000700006b8c000]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
        at java.net.SocketInputStream.read(SocketInputStream.java:171)
        at java.net.SocketInputStream.read(SocketInputStream.java:141)
        at org.apache.http.impl.conn.LoggingInputStream.read(LoggingInputStream.java:84)
        at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
        at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
        at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
        at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
        at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
        at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
        at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
        at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:157)
        at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
        at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
        at com.sap.cloud.security.xsuaa.client.DefaultOAuth2TokenService.executeRequest(DefaultOAuth2TokenService.java:84)
        at com.sap.cloud.security.xsuaa.client.DefaultOAuth2TokenService.requestAccessToken(DefaultOAuth2TokenService.java:78)
        at com.sap.cloud.security.xsuaa.client.AbstractOAuth2TokenService.getAndCacheToken(AbstractOAuth2TokenService.java:305)
        at com.sap.cloud.security.xsuaa.client.AbstractOAuth2TokenService.getOrRequestAccessToken(AbstractOAuth2TokenService.java:263)
        at com.sap.cloud.security.xsuaa.client.AbstractOAuth2TokenService.getOAuth2TokenResponse(AbstractOAuth2TokenService.java:252)
        at com.sap.cloud.security.xsuaa.client.AbstractOAuth2TokenService.retrieveAccessTokenViaClientCredentialsGrant(AbstractOAuth2TokenService.java:112)
        at com.sap.cloud.security.xsuaa.tokenflows.ClientCredentialsTokenFlow.requestTechnicalUserToken(ClientCredentialsTokenFlow.java:183)
        at com.sap.cloud.security.xsuaa.tokenflows.ClientCredentialsTokenFlow.execute(ClientCredentialsTokenFlow.java:141)
        at org.cloudfoundry.promregator.auth.OAuth2XSUAAEnricher.getJWT(OAuth2XSUAAEnricher.java:86)
        at org.cloudfoundry.promregator.auth.OAuth2XSUAAEnricher.getJWTAndCatchExceptions(OAuth2XSUAAEnricher.java:76)
        at org.cloudfoundry.promregator.auth.OAuth2XSUAAEnricher.enrichWithAuthentication(OAuth2XSUAAEnricher.java:65)
        at org.cloudfoundry.promregator.auth.OAuth2XSUAAEnricherTest.testBadJWTCall(OAuth2XSUAAEnricherTest.java:79)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$105/333362446.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall$$Lambda$106/1501587365.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$$Lambda$248/385337537.apply(Unknown Source)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$269/1117509763.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$192/2095490653.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$191/1253946629.invoke(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$190/951880373.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$196/1121453612.accept(Unknown Source)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$192/2095490653.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$191/1253946629.invoke(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$190/951880373.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$196/1121453612.accept(Unknown Source)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$192/2095490653.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$191/1253946629.invoke(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$190/951880373.execute(Unknown Source)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator$$Lambda$153/418304857.accept(Unknown Source)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:84)
        at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
        at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

During this state I'm able to call the mock handler on the server via

```
curl -vvvv  -X POST http://localhost:9001/oauth/token  --header "Content-Type: application/x-www-form-urlencoded" --data "grant_type=client_credentials&client_secret=client_secret&client_id=client_id"
```

I get the exected response, including the body:

```
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 9001 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9001 (#0)
> POST /oauth/token HTTP/1.1
> Host: localhost:9001
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 77
> 
* upload completely sent off: 77 out of 77 bytes
< HTTP/1.1 401 Unauthorized
< Date: Thu, 18 Nov 2021 11:34:18 GMT
< Content-length: 62
< 
* Connection #0 to host localhost left intact
{"error":"unauthorized","error_description":"Bad credentials"}* Closing connection 0
```

I also tried to configure the http client in our factory by providing a request config:

```
RequestConfig config = RequestConfig.custom()
					  .setConnectTimeout(1 * 1000)
					  .setConnectionRequestTimeout(1 * 1000)
					  .setSocketTimeout(1 * 1000).build();
```

This seems not to have an impact.


I also thought about a general mis-configuration of the http client which might try to keep the connection open for some time which might also affect the real-world and not only the tests. So I disabled the token client test and reworked the code so that several attempts for token fetching are executed. I was able to see that each request ended up in a new token. So for the real-world everything looks fine. I can say that in the real world we are able to fetch several tokens. It is not like only the first attempt works.

I have no idea how to solve this. @eaglerainbow : maybe you could have a look since you might be more experienced with the mock http server setup used for the tests.

Reply:
> There is a second request triggered, but that never hits the DefaultOAuthHttpHandler as we can see since the log line log.debug("Request was received at handler"); is not issued to the log.

Although all this still this confuses me. Even if you do not close the HTTP(S) connection, you should be able to fire two requests. And hence, if two requests are triggered, also the mock server should hit `DefaultOAuthHttpHandler`: The server can't know about caching data...
This is nothing we should bother too much about, but still, if I find some time, I'll have a look into this...

Reply:
I debugged a little around here and it seems that `token-client` rightly tries to use HTTP persistent connections, but the jdk's [simple HttpServer](https://docs.oracle.com/javase/9/docs/api/com/sun/net/httpserver/package-summary.html#) has some issues with that:

```
22:49:54.407 [main] DEBUG com.sap.cloud.security.xsuaa.client.DefaultOAuth2TokenService - Requesting access token from url http://localhost:9001/oauth/token with headers [Content-Type: application/x-www-form-urlencoded, Accept: application/json]
22:51:01.678 [main] DEBUG org.apache.http.client.protocol.RequestAddCookies - CookieSpec selected: default
22:51:01.679 [main] DEBUG org.apache.http.client.protocol.RequestAuthCache - Auth cache not set in the context
22:51:01.681 [main] DEBUG org.apache.http.impl.conn.PoolingHttpClientConnectionManager - Connection request: [route: {}->http://localhost:9001][total available: 1; route allocated: 1 of 2; total allocated: 1 of 20]
22:51:01.685 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "[read] I/O error: Read timed out"
22:51:01.688 [main] DEBUG org.apache.http.impl.conn.PoolingHttpClientConnectionManager - Connection leased: [id: 0][route: {}->http://localhost:9001][total available: 0; route allocated: 1 of 2; total allocated: 1 of 20]
22:51:01.688 [main] DEBUG org.apache.http.impl.conn.DefaultManagedHttpClientConnection - http-outgoing-0: set socket timeout to 0
22:51:01.689 [main] DEBUG org.apache.http.impl.execchain.MainClientExec - Executing request POST /oauth/token HTTP/1.1
22:51:01.690 [main] DEBUG org.apache.http.impl.execchain.MainClientExec - Target auth state: UNCHALLENGED
22:51:01.691 [main] DEBUG org.apache.http.impl.execchain.MainClientExec - Proxy auth state: UNCHALLENGED
```

There are two things that confuses me:  First thing is this
``` 
22:51:01.685 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "[read] I/O error: Read timed out"
```
which means that the client fails on reading data from the server, i.e. the client expects the (mock) server still to write some data to the queue to be in sync.

The second thing which confuses me is that I debugged into the waiting HTTP Server, which is stuck in 

``` 
WindowsSelectorImpl.doSelect(Consumer<SelectionKey>, long) line: 182	
```
even if the second request on the persistent connection comes in. Looking into the logic there a little, it seems that the server isn't `select()`ing on the persistent connection anymore and hence OS-polling does not retrigger the main loop of the webservice.

Long story short: It seems that it's the mock server that is failing/buggy/misconfigured/....
Your `Connection: close` was the bazooka, as you force the server to close the connection upon completing the request, releasing the resources in your client connection pool. On the next request, the client is forced to establish a new connection where then things start from scratch. 

Now, we need to think about whether changing productive coding for making unit tests run is the right thing here. I would assume that the webdispatchers of CF platforms may handle persistent connections properly. Otherwise they would also get major issues in other situations (not just fetching XSUAA JWTs). So, client-rejecting persistent connections is not really necessary. 
However, on the other hand, how likely is it that we may benefit from persistent connections? Our requests are not like loading a webpage in a browser (where hundreds of requests for a single webpage might need to be triggered), but the number of requests be very limited - and we will only request it again shortly before the JWT expires (which may easily be in the order of hours). Until that point in time the persistent connection will be long-gone anyways. 

The only situation where we *might* benefit from persistent connections would be, if we had multiple different OAuth2xsuaa* authenticators configured. Especially during or shortly after startup of Promregator we will fetch all of these JWTs in a very short timeframe. As connection pooling would happen on HTTPClient's level, we would be able to reuse persistent connections for multiple different requests, hence saving resources at routers and the platform. However, I would expect that this benefit can only be observed if a large number of authenticators (say: hundreds of them) were registered. I have never seen such a configuration. Instead, my expectation would be that we will run into other, more severe scaling issues at various other places than latency issues due to "high number of non-persistent HTTP requests".

Having said that, I think your 306dcead0c1e461ed4e532de18e6a378f85b70a6 is the right way to go right now.